### DISCUSSION TITLE
Method `graphql_sync` returns success=True although there is errors

### DISCUSSION THREAD

**Discussion by mahsa-lotfi92** [mirumee/ariadne#543] [Q&A]
Hello everyone. I have a question regarding the graphql_sync method.

When I try this query:

```graphql
query{
    MyQuery (data: {
          code: 1,
    }){
      result
    }
  }
```
the method `graphql_sync` returns `(False, {'errors': [{'message': 'String cannot represent a non string value: 1', 'locations': [{'line': 3, 'column': 25}], 'path': None}]}` which is right, because the code needs to be a string. But when I try to  use variables like this:

```
query($data: MyQueryInput!) {
      MyQuery(data: $data) {
            result
      }
    }
```
with variables like this:
```
"data": {
        "code": 1
    }
```
the method `graphql_sync` returns `(True, {'data': None, 'errors': [{'message': "Variable '$data' got invalid value 1 at 'data.code'; String cannot represent a non string value: 1", 'locations': [{'line': 1, 'column': 7}], 'path': None}]}`

Why this behaves differently? I wanted to use the success flag to change the status code of response like it said here:
https://ariadne.readthedocs.io/en/0.3.0/custom-server.html. But this different behaviour makes it inconsistent. Did I get something wrong here?

Many thanks for the help.

**Comment by rafalp** [comment]
> Did I get something wrong here?

Generally speaking for GraphQL servers its accepted that HTTP code different than 200 means that query wasn't executed at all, not that it was executed with errors.

HTTP status codes aren't used by in GraphQL client/server communication, and this is reflected by API design of GraphQL client libs in wild. Don't rely on them in your communication implementation or you'll end fighting the stack.

There's also a separate discussion on relying on `errors` key from GraphQL result for communicating errors in business logic from app to client. It was considered [good idea back in 2018](https://www.apollographql.com/blog/full-stack-error-handling-with-graphql-apollo-5c12da407210/), but these days its considered more of an antipattern, with [intermediate result type being preferable](https://ariadnegraphql.org/docs/error-messaging). Biggest advantage of this approach is developer control on how application specific errors are reported and clean difference between `result.errors` telling you when developer did something wrong vs `result.data.someMutation.errors` telling you when client did something wrong.

**Reply by mahsa-lotfi92** [reply]
Thank you so much for the quick and complete response. I got it I think. I am new to graphql too and I need to investigate more.
